test/hotspot/jtreg/testlibrary_tests/ir_framework/tests/TestNotCompilable.java

```diff
@@ -0,0 +1,202 @@
+        runWithExcludeExpectFailure(TestClassA.class, noWarmup);
+        runWithExcludeExpectFailure(TestClassB.class, noWarmup);
+        // Note: @Run does not fail, but the @IR detects that there is no compilation, and fails.
+        runWithExcludeExpectFailure(TestClassE.class, noWarmup);
+        runOptoNoExecuteExpectFailure(TestClassA.class, noWarmup);
+        runOptoNoExecuteExpectFailure(TestClassB.class, noWarmup);
+
+        // Forbid compilation -> annotation allows not compilable -> should pass.
+        runWithExcludeExpectSuccess(TestClassC.class, noWarmup);
+        runWithExcludeExpectSuccess(TestClassD.class, noWarmup);
+        runWithExcludeExpectSuccess(TestClassF.class, noWarmup);
+        runOptoNoExecuteExpectSuccess(TestClassC.class, noWarmup);
+        runOptoNoExecuteExpectSuccess(TestClassD.class, noWarmup);
+        // Note: @Run does not fail because of missing compilation. And OptoNoExecute does still
+        //       print IR before it bails out, and we can successfully match it.
+        runOptoNoExecuteExpectSuccess(TestClassE.class, noWarmup);
+        runOptoNoExecuteExpectSuccess(TestClassF.class, noWarmup);
+
+        // Forbid compilation, but allow methods not to compile -> should pass.
+        runWithExcludeAndGlobalAllowNotCompilable(TestClassA.class, noWarmup);
+        runWithExcludeAndGlobalAllowNotCompilable(TestClassB.class, noWarmup);
+        runWithExcludeAndGlobalAllowNotCompilable(TestClassC.class, noWarmup);
+        runWithExcludeAndGlobalAllowNotCompilable(TestClassD.class, noWarmup);
+        runWithExcludeAndGlobalAllowNotCompilable(TestClassE.class, noWarmup);
+        runWithExcludeAndGlobalAllowNotCompilable(TestClassF.class, noWarmup);
+        runOptoNoExecuteAndGlobalAllowNotCompilable(TestClassA.class, noWarmup);
+        runOptoNoExecuteAndGlobalAllowNotCompilable(TestClassB.class, noWarmup);
+        runOptoNoExecuteAndGlobalAllowNotCompilable(TestClassC.class, noWarmup);
+        runOptoNoExecuteAndGlobalAllowNotCompilable(TestClassD.class, noWarmup);
+        runOptoNoExecuteAndGlobalAllowNotCompilable(TestClassE.class, noWarmup);
+        runOptoNoExecuteAndGlobalAllowNotCompilable(TestClassF.class, noWarmup);
+    }
+
+    private static void runNormal(Class c, boolean noWarmup) {
+        TestFramework framework = new TestFramework(c);
+        if (noWarmup) { framework.setDefaultWarmup(0); }
+        framework.start();
+    }
+
+    private static void runWithExcludeExpectFailure(Class c, boolean noWarmup) {
+        TestFramework framework = new TestFramework(c);
+        framework.addFlags("-XX:CompileCommand=exclude,*TestClass*::test*");
+        if (noWarmup) { framework.setDefaultWarmup(0); }
+        try {
+            framework.start();
+            throw new RuntimeException("should have thrown TestRunException");
+        } catch (TestVMException e) {
+        } catch (IRViolationException e) {}
```

Comment:
This is a mismatch. Which exceptions are now expected? Same below.

Reply:
Adjusted and added comment :)